src/background_thread.c

```diff
@@ -177,13 +176,21 @@ background_work_sleep_once(tsdn_t *tsdn, background_thread_info_t *info,
     unsigned ind) {
 	uint64_t ns_until_deferred = BACKGROUND_THREAD_DEFERRED_MAX;
 	unsigned narenas = narenas_total_get();
+	bool slept_indefinitely = background_thread_indefinite_sleep(info);
 
 	for (unsigned i = ind; i < narenas; i += max_background_threads) {
 		arena_t *arena = arena_get(tsdn, i, false);
 		if (!arena) {
 			continue;
 		}
-		arena_do_deferred_work(tsdn, arena);
+		/*
```

Comment:
I don't totally follow the reasoning here; why should we skip work depending on the conditions of the wakeup?

Reply:
We wake up a thread from indefinite sleep when we generate deferred work. At this point we've just generated the work and it doesn't need to be done instantly. We wake up the thread from the indefinite sleep and put it into another sleep, this time for the period of deferral.

Here is a hypothetical example:
1. Thread sleeps indefinitely
2. HPA generated deferred work in 5 sec
3. We wake up the thread, it checks that the work should be done in 5 sec
4. Thread sleeps for another 5 sec
5. In 5 sec the thread wakes up and does the work
6. Thread checks if there's still deferred work, then it sleeps for `time_until_deferred_work`. Otherwise, it sleeps indefinitely.

If the thread did all the work instantly, in case of HPA if would keep flipping between indefinite sleep and immediate work execution.

Reply:
Ahh, so just to confirm my understanding: when a background thread's sleep time gets shortened, it wakes up twice: the first time to sleep with a new parameter, and the second time to actually do the work?

Reply:
Almost. It wakes up twice only when sleep time gets shortened and it is indefinite. Otherwise it doesn't wake up twice, which is a limitation of current implementation.